src/Enum.php

```diff
@@ -64,157 +60,237 @@ public static function __callStatic($name, $arguments)
+    public function __toString(): string
     {
-        if (strlen($name) > 2 && strpos($name, 'is') === 0) {
-            return $this->equals(substr($name, 2));
-        }
-
-        if (isset(self::resolve()[strtolower($name)])) {
-            return static::__callStatic($name, $arguments);
-        }
-
-        throw new \BadMethodCallException(sprintf('Call to undefined method %s->%s()', static::class, $name));
+        return $this->getValue();
     }
 
-    /**
-     * @param string|\Spatie\Enum\Enum $enum
-     *
-     * @return bool
-     */
-    public function equals($enum): bool
+    public function getIndex(): int
     {
-        if (is_string($enum)) {
-            $enum = static::from($enum);
-        }
+        return $this->index;
+    }
 
-        if (! $enum instanceof $this) {
-            return false;
-        }
+    public static function getIndices(): array
+    {
+        return array_column(static::resolve(), 'index');
+    }
 
-        if ($enum->value !== $this->value) {
-            return false;
-        }
+    public function getValue(): string
+    {
+        return $this->value;
+    }
 
-        return true;
+    public static function getValues(): array
+    {
+        return array_column(static::resolve(), 'value');
     }
 
-    /**
-     * @param string[]|\Spatie\Enum\Enum[] $enums
-     *
-     * @return bool
-     */
-    public function isOneOf(array $enums): bool
+    public function isAny(array $values): bool
     {
-        foreach ($enums as $enum) {
-            if ($this->equals($enum)) {
+        foreach ($values as $value) {
+            if ($this->isEqual($value)) {
                 return true;
             }
         }
 
         return false;
     }
 
-    public function __toString(): string
+    public function isEqual($value): bool
     {
-        return $this->value;
+        if (is_int($value) || is_string($value)) {
+            $enum = static::make($value);
+        } elseif ($value instanceof $this) {
+            $enum = $value;
+        }
+
+        if (
+            isset($enum)
+            && $enum instanceof $this
+            && $enum->getValue() === $this->getValue()
+        ) {
+            return true;
+        }
+
+        return false;
     }
 
     public function jsonSerialize()
     {
-        return $this->value;
+        return $this->getValue();
+    }
+
+    public static function make($value): Enumerable
+    {
+        if (! (is_int($value) || is_string($value))) {
+            throw new TypeError(sprintf('%s::make() expects string|int as argument but %s given', static::class, gettype($value)));
+        }
+
+        $name = null;
+        $index = null;
+
+        if (is_int($value)) {
+            if (! static::isValidIndex($value)) {
+                throw new InvalidIndexException($value, static::class);
+            }
+
+            $name = array_combine(static::getIndices(), array_keys(static::resolve()))[$value];
```

Comment:
Any change we can cache these array_combine calls so that this hasn't be done very time an enum is constructed?

Reply:
Yeah, will try to add some more poor-man caches to keep these values.

Reply:
Hey, adding the caches has no impact on performance. I've added a speed test and the uncached `make()` call needs the same time like the cached one.
And adding two static variables and two methods for a single call without any performance impact isn't a good trade!?

> Spatie\Enum\OldTests\WeekDayEnumSpeedTest::can_make_faster_from_cache_than_build
> Failed asserting that 1.4 is less than 1.4.

> Spatie\Enum\OldTests\WeekDayEnumSpeedTest::can_load_stable_from_cache
> Failed asserting that 0.4 is less than 0.38000000000000006.

> Spatie\Enum\OldTests\WeekDayEnumSpeedTest::can_make_faster_from_cache_than_build
> Failed asserting that 0.6 is less than 0.6.

Reply:
You're right, let's keep it as is